# <h1><center>**A Tale of Two cities - Clustering the Neighbourhoods of London and Paris**</center></h1>

**Authored By:** Venkata Shanmukeswar Tata <br>

**Date:** 15th October 2020<br>

## **1. Introduction**<br>
A Tale of Two Cities was published in 1859, right between Little Dorrit and Great Expectations, two of other Dicken's novels. These are novels set in Dicken's own century, if several decades before the years of their publication. And they're focused on individual stories of crime, poverty and class in 19th Century England and Europe. Tale of Two Cities, by contrast, is one of only two works Dickens wrote, that could properly be considered historical novels. The other is Barnaby Rudge, set in the 1770s and the early 1780s. 
<br><br>

A Tale of Two Cities has always been considered one of the least Dickensian of Dickens' novels. For example, a scholar writing for the Cambridge History of English Literature, published in 1941, wrote, many people who do not care for the rest of Dickens like it greatly; many who are enthusiastic about Dickens refuse to give it a second reading. 
<br><br>

It is the least Dickensian of all the tales. And it does have an odd relationship to his corpeous. One reason, perhaps, is that the novel feels somehow less Victorian, less sentimental than other of Dickens' well-loved novels. Dickens, himself, writing about a political controversy taking place while he was working on the novel, gives us a clue about these differences. When the great remonstrance came out, he writes, I was in the thick of my story, and was always busy with it.

<br><br>
London and Paris are quite the popular tourist and vacation destinations for people all around the world. They are diverse and multicultural and offer a wide variety of experiences that is widely sought after. We try to group the neighbourhoods of London and Paris respectively and draw insights to what they look like now.

## **2. Business Problem**<br>

The aim is to help tourists choose their destinations depending on the experiences that the neighbourhoods have to offer and what they would want to have. This also helps people make decisions if they are thinking about migrating to London or Paris or even if they want to relocate neighbourhoods within the city. Our findings will help stakeholders make informed decisions and address any concerns they have including the different kinds of cuisines, provision stores and what the city has to offer.

Our goal is to perform a comparison of the two cities to see how similar or dissimilar they are. Such techniques allow users to identify similar neighbourhoods among cities based on amenities or services being offered locally, and thus can help in understanding the local area activities, what are the hubs of different activities, how citizens are experiencing the city, and how they are utilising its resources.
What kind of clientele would benefit from such an analysis?
1.	A potential job seeker with transferable skills may wish to search for jobs in selective cities which provide the most suitable match for their qualifications and experience in terms of salaries, social benefits, or even in terms of a culture fit for expats.
2.	Further, a person buying or renting a home in a new city may want to look for recommendations for locations in the city similar to other cities known to them.
3.	Similarly, a large corporation looking to expand its locations to other cities might benefit from such an analysis.
4.	Many within-city urban planning computations might also benefit from modelling a city’s relationship to other cities.

## **3. Data Description**<br>
We require geographical location data for both London and Paris. Postal codes in each city serve as a starting point. Using Postal codes we use can find out the neighborhoods, boroughs, venues and their most popular venue categories.

### **3.1 London**<br>
To derive our solution, We scrape our data from https://en.wikipedia.org/wiki/List_of_areas_of_London<br>

This wikipedia page has information about all the neighbourhoods, we limit it London.

1. borough : Name of Neighbourhood
2. town : Name of borough
3. post_code : Postal codes for London.<br>

This wikipedia page lacks information about the geographical locations. To solve this problem we use ArcGIS API

### **3.2 ArcGIS API**<br>
ArcGIS Online enables you to connect people, locations, and data using interactive maps. Work with smart, data-driven styles and intuitive analysis tools that deliver location intelligence. Share your insights with the world or specific groups.<br>

More specifically, we use ArcGIS to get the geo locations of the neighbourhoods of London. The following columns are added to our initial dataset which prepares our data.

1. latitude : Latitude for Neighbourhood
2. longitude : Longitude for Neighbourhood

### **3.3 Paris**<br>

To derive our solution, We leverage JSON data available at https://www.data.gouv.fr/fr/datasets/r/e88c6fda-1d09-42a0-a069-606d3259114e

The JSON file has data about all the neighbourhoods in France, we limit it to Paris.

1. postal_code : Postal codes for France
2. nom_comm : Name of Neighbourhoods in France
3. nom_dept : Name of the boroughs, equivalent to towns in France
4. geo_point_2d : Tuple containing the latitude and longitude of the Neighbourhoods.

### **3.4 Foursquare API Data**<br>
We will need data about different venues in different neighbourhoods of that specific borough. In order to gain that information we will use "Foursquare" locational information. Foursquare is a location data provider with information about all manner of venues and events within an area of interest. Such information includes venue names, locations, menus and even photos. As such, the foursquare location platform will be used as the sole data source since all the stated required information can be obtained through the API.

After finding the list of neighbourhoods, we then connect to the Foursquare API to gather information about venues inside each and every neighbourhood. For each neighbourhood, we have chosen the radius to be 500 meters.

The data retrieved from Foursquare contained information of venues within a specified distance of the longitude and latitude of the postcodes. The information obtained per venue as follows:

1. Neighbourhood : Name of the Neighbourhood
2. Neighbourhood Latitude : Latitude of the Neighbourhood
3. Neighbourhood Longitude : Longitude of the Neighbourhood
4. Venue : Name of the Venue
5. Venue Latitude : Latitude of Venue
6. Venue Longitude : Longitude of Venue
7. Venue Category : Category of Venue<br>

Based on all the information collected for both London and Paris, we have sufficient data to build our model. We cluster the neighbourhoods together based on similar venue categories. We then present our observations and findings. Using this data, our stakeholders can take the necessary decision.

## **4. Methodology**<br>

In [ ]:
import pandas as pd
import requests
import numpy as np
import matplotlib.cm as cm
import matplotlib.colors as colors
import folium
from sklearn.cluster import KMeans

We will be creating our model with the help of Python so we start off by importing all the required packages.

> import pandas as pd<br>
> import requests<br>
> import numpy as np<br>
> import matplotlib.cm as cm<br>
> import matplotlib.colors as colors<br>
> import folium<br>
> from sklearn.cluster import KMeans<br>


Package breakdown:

1. Pandas : To collect and manipulate data in JSON and HTMl and then data analysis
2. requests : Handle http requests
3. matplotlib : Detailing the generated maps
4. folium : Generating maps of London and Paris
5. sklearn : To import Kmeans which is the machine learning model that we are using.<br>

The approach taken here is to explore each of the cities individually, plot the map to show the neighbourhoods being considered and then build our model by clustering all of the similar neighbourhoods together and finally plot the new map with the clustered neighbourhoods. We draw insights and then compare and discuss our findings.

### **4.1 Data Collection**<br>
In the data collection stage, we begin with collecting the required data for the cities of London and Paris. We need data that has the postal codes, neighbourhoods and boroughs specific to each of the cities.<br>

To collect data for London, we scrape the List of areas of London wikipedia page to take the 2nd table using the following code:

In [ ]:
url_london = "https://en.wikipedia.org/wiki/List_of_areas_of_London"
wiki_london_url = requests.get(url_london)
wiki_london_data = pd.read_html(wiki_london_url.text)
wiki_london_data = wiki_london_data[1]
wiki_london_data

Location                     London borough  ... Dial code OS grid ref
0         Abbey Wood              Bexley, Greenwich [7]  ...       020    TQ465785
1              Acton  Ealing, Hammersmith and Fulham[8]  ...       020    TQ205805
2          Addington                         Croydon[8]  ...       020    TQ375645
3         Addiscombe                         Croydon[8]  ...       020    TQ345665
4        Albany Park                             Bexley  ...       020    TQ478728
..               ...                                ...  ...       ...         ...
528         Woolwich                          Greenwich  ...       020    TQ435795
529   Worcester Park       Sutton, Kingston upon Thames  ...       020    TQ225655
530  Wormwood Scrubs             Hammersmith and Fulham  ...       020    TQ225815
531          Yeading                         Hillingdon  ...       020    TQ115825
532         Yiewsley                         Hillingdon  ...       020    TQ063804

[533 rows x 6 columns]

To collect data for Paris, we download the JSON file containg all the postal codes of France from https://www.data.gouv.fr/fr/datasets/r/e88c6fda-1d09-42a0-a069-606d3259114e<br>

Using Pandas we load the table after reading the JSON file:

In [ ]:
!wget -q -O 'france-data.json' https://www.data.gouv.fr/fr/datasets/r/e88c6fda-1d09-42a0-a069-606d3259114e
print("Data Downloaded!")
paris_raw = pd.read_json('france-data.json')
paris_raw.head()

Data Downloaded!


datasetid  ...               record_timestamp
0  correspondances-code-insee-code-postal  ...  2016-09-21T00:29:06.175+02:00
1  correspondances-code-insee-code-postal  ...  2016-09-21T00:29:06.175+02:00
2  correspondances-code-insee-code-postal  ...  2016-09-21T00:29:06.175+02:00
3  correspondances-code-insee-code-postal  ...  2016-09-21T00:29:06.175+02:00
4  correspondances-code-insee-code-postal  ...  2016-09-21T00:29:06.175+02:00

[5 rows x 5 columns]

### **4.2 Data preprocessing**<br>

For London, We replace the spaces with underscores in the title.The borough column has numbers within square brackets that we remove using:

In [ ]:
wiki_london_data.rename(columns=lambda x: x.strip().replace(" ", "_"), inplace=True)
wiki_london_data['London\xa0borough'] = wiki_london_data['London\xa0borough'].map(lambda x: x.rstrip(']').rstrip('0123456789').rstrip('['))

For Paris, we break down each of the nested fields and create the dataframe that we need:

In [ ]:
paris_field_data = pd.DataFrame()
for f in paris_raw.fields:
    dict_new = f
    paris_field_data = paris_field_data.append(dict_new, ignore_index=True)

paris_field_data.head()

code_arr code_cant code_comm  ...            statut superficie z_moyen
0        3        03       645  ...    Commune simple      999.0   121.0
1        3        20       133  ...    Commune simple     1082.0    88.0
2        1        09       378  ...    Commune simple      313.0   150.0
3        5        14       243  ...  Chef-lieu canton      579.0    71.0
4        3        34       003  ...  Chef-lieu canton      232.0    70.0

[5 rows x 17 columns]

### **4.3 Feature Selection**<br>

For both of our datasets, we need only the borough, neighbourhood, postal codes and geolocations (latitude and longitude). So we end up selecting the columns that we need by:

In [ ]:
df1 = wiki_london_data.drop( [ wiki_london_data.columns[0], wiki_london_data.columns[4], wiki_london_data.columns[5] ], axis=1)

df_2 = paris_field_data[['postal_code','nom_comm','nom_dept','geo_point_2d']]

### **4.4 Feature Engineering**<br>

Both of our Datasets actually contain information related to all the cities in the country. We can narrow down and further process the data by selecting only the neighbourhoods pertaining to 'London' and 'Paris'

In [ ]:
df1 = df1[df1['Post_town'].str.contains('LONDON')]

df_paris = df_2[df_2['nom_dept'].str.contains('PARIS')].reset_index(drop=True)

Looking over our London dataset, we can see that we don't have the geolocation data. We need to extrapolate the missing data for our neighbourhoods. We perform this by leveraging the ArcGIS API. With the Help of ArcGIS API we can get the latitude and longitude of our London neighbourhood data.

In [ ]:
from arcgis.geocoding import geocode
from arcgis.gis import GIS
gis = GIS()

Defining London arcgis geocode function to return latitude and longitude

In [ ]:
def get_x_y_uk(address1):
   lat_coords = 0
   lng_coords = 0
   g = geocode(address='{}, London, England, GBR'.format(address1))[0]
   lng_coords = g['location']['x']
   lat_coords = g['location']['y']
   return str(lat_coords) +","+ str(lng_coords)

Passing postal codes of london to get the geographical co-ordinates

In [ ]:
geo_coordinates_uk = df1['Postcode\xa0district']    
geo_coordinates_uk
coordinates_latlng_uk = geo_coordinates_uk.apply(lambda x: get_x_y_uk(x))

Extracting the latitude from our previously collected coordinates

In [ ]:
lat_uk = coordinates_latlng_uk.apply(lambda x: x.split(',')[0])
lat_uk

0      51.492450000000076
1       51.51324000000005
6       51.51200000000006
7       51.51651000000004
9       51.41009000000008
              ...        
523    51.589770000000044
524     51.50642000000005
527    51.615920000000074
528     51.48207000000008
530     51.50645000000003
Name: Postcode district, Length: 310, dtype: object

Extracting the Longitude from our previously collected coordinates

In [ ]:
lng_uk = coordinates_latlng_uk.apply(lambda x: x.split(',')[1])
lng_uk

0       0.12127000000003818
1       -0.2674599999999714
6      -0.08057999999994081
7      -0.11967999999995982
9      -0.05682999999993399
               ...         
523    0.030520000000024083
524     -0.1272099999999341
527     -0.1767399999999384
528     0.07143000000002075
530     -0.2369099999999662
Name: Postcode district, Length: 310, dtype: object

We proceed with Merging our source data with the geographical co-ordinates to make our dataset ready for the next stage

In [ ]:
london_merged = pd.concat([df1,lat_uk.astype(float), lng_uk.astype(float)], axis=1)
london_merged.columns= ['borough','town','post_code','latitude','longitude']
london_merged

borough  ... longitude
0                Bexley, Greenwich   ...   0.12127
1    Ealing, Hammersmith and Fulham  ...  -0.26746
6                              City  ...  -0.08058
7                       Westminster  ...  -0.11968
9                           Bromley  ...  -0.05683
..                              ...  ...       ...
523                       Redbridge  ...   0.03052
524       Redbridge, Waltham Forest  ...  -0.12721
527                          Barnet  ...  -0.17674
528                       Greenwich  ...   0.07143
530          Hammersmith and Fulham  ...  -0.23691

[310 rows x 5 columns]

As for our Paris dataset, we don't need to get the geo coordinates using an external data source or collect it with the ArcGIS API call since we already have it stored in the geo_point_2d column as a tuple in the df_paris dataframe.

We just need to extract the latitude and longitude for the column:



In [ ]:
paris_latlng = df_paris['geo_point_2d'].astype('str')

paris_lat = paris_latlng.apply(lambda x: x.split(',')[0])
paris_lat = paris_lat.apply(lambda x: x.lstrip('['))

paris_lng = paris_latlng.apply(lambda x: x.split(',')[1])
paris_lng = paris_lng.apply(lambda x: x.rstrip(']'))

paris_geo_lat  = pd.DataFrame(paris_lat.astype(float))
paris_geo_lat.columns=['Latitude']

paris_geo_lng = pd.DataFrame(paris_lng.astype(float))
paris_geo_lng.columns=['Longitude']

We then create our Paris dataset with the required information:

In [ ]:
paris_combined_data = pd.concat([df_paris.drop('geo_point_2d', axis=1), paris_geo_lat, paris_geo_lng], axis=1)
paris_combined_data

postal_code                  nom_comm nom_dept   Latitude  Longitude
0        75009   PARIS-9E-ARRONDISSEMENT    PARIS  48.876896   2.337460
1        75002   PARIS-2E-ARRONDISSEMENT    PARIS  48.867903   2.344107
2        75011  PARIS-11E-ARRONDISSEMENT    PARIS  48.859415   2.378741
3        75003   PARIS-3E-ARRONDISSEMENT    PARIS  48.863054   2.359361
4        75006   PARIS-6E-ARRONDISSEMENT    PARIS  48.848968   2.332671
5        75019  PARIS-19E-ARRONDISSEMENT    PARIS  48.886869   2.384694
6        75020  PARIS-20E-ARRONDISSEMENT    PARIS  48.863187   2.400820
7        75015  PARIS-15E-ARRONDISSEMENT    PARIS  48.840155   2.293559
8        75005   PARIS-5E-ARRONDISSEMENT    PARIS  48.844509   2.349859
9        75007   PARIS-7E-ARRONDISSEMENT    PARIS  48.856083   2.312439
10       75008   PARIS-8E-ARRONDISSEMENT    PARIS  48.872527   2.312583
11       75013  PARIS-13E-ARRONDISSEMENT    PARIS  48.828718   2.362468
12       75012  PARIS-12E-ARRONDISSEMENT    PARIS  48.835156   2.419807
13       75018  PARIS-18E-ARRONDISSEMENT    PARIS  48.892735   2.348712
14       75001  PARIS-1ER-ARRONDISSEMENT    PARIS  48.862630   2.336293
15       75010  PARIS-10E-ARRONDISSEMENT    PARIS  48.876029   2.361113
16       75016  PARIS-16E-ARRONDISSEMENT    PARIS  48.860399   2.262100
17       75004   PARIS-4E-ARRONDISSEMENT    PARIS  48.854228   2.357362
18       75017  PARIS-17E-ARRONDISSEMENT    PARIS  48.887337   2.307486
19       75014  PARIS-14E-ARRONDISSEMENT    PARIS  48.828993   2.327101

### **4.5 Visualizing the neighbourhoods of london and paris**<br>

**4.5.1 London**<br>

Getting the geocode for London to help visualize it on the map

In [ ]:
london = geocode(address='London, England, GBR')[0]
london_lng_coords = london['location']['x']
london_lat_coords = london['location']['y']

In [ ]:
london_lat_coords

51.50642000000005

In [ ]:
london_lng_coords

-0.1272099999999341

To help visualize the Map of London and the neighbourhoods in London, we make use of the folium package.

In [ ]:
# Creating the map of London
map_London = folium.Map(location=[london_lat_coords, london_lng_coords], zoom_start=12)
map_London

# adding markers to map
for latitude, longitude, borough, town in zip(london_merged['latitude'], london_merged['longitude'], london_merged['borough'], london_merged['town']):
    label = '{}, {}'.format(town, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [latitude, longitude],
        radius=5,
        popup=label,
        color='red',
        fill=True
        ).add_to(map_London)  
    
map_London

To proceed with the next part, we need to define Foursquare API credentials.<br>

Using Foursquare API, we are able to get the venue and venue categories around each neighbourhood in London.

In [ ]:
CLIENT_ID = 'CP3S15XNMLK4PPYDXZGL2A1LAPYBX21WHD3RYY24DAHX2AVI' 
CLIENT_SECRET = 'PL2YHWGJYR1YILZ0GHNPP1EFY3OQCGNFBJTNVKV5RSNUXW1T'
VERSION = '20201016' # Foursquare API version

Defining a function to get the neraby venues in the neighbourhood. This will help us get venue categories which is important for our analysis

In [ ]:
LIMIT=100
def getNearbyVenues(names, latitudes, longitudes, radius=500):
  venues_list = []
  for name, lat, lng in zip(names, latitudes, longitudes):
    print(name)

    # create the API request URL
    url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
        CLIENT_ID,
        CLIENT_SECRET,
        VERSION,
        lat,
        lng,
        radius,
        LIMIT
    )

    # make the GET request
    results = requests.get(url).json()["response"]['groups'][0]['items']

    # return only relevant information for each nearby venue
    venues_list.append([(name, lat, lng, v['venue']['name'], v['venue']['categories'][0]['name']) for v in results])
    
  nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
  nearby_venues.columns = ['Neighbourhood', 'Neighbourhood Latitude','Neighbourhood Longitude','Venue','Venue Category']
  return (nearby_venues)

Getting the venues in London

In [ ]:
venues_in_London = getNearbyVenues(london_merged['borough'], london_merged['latitude'], london_merged['longitude'])

Bexley, Greenwich 
Ealing, Hammersmith and Fulham
City
Westminster
Bromley
Islington
Islington
Barnet
Enfield
Wandsworth
Southwark
City
Richmond upon Thames
Barnet
Islington
Wandsworth
Westminster
Bromley
Newham
Ealing
Westminster
Lewisham
Camden
Southwark
Tower Hamlets
Bexley
City
Lewisham
Greenwich
Tower Hamlets
Camden
Haringey
Tower Hamlets
Haringey
Barnet
Brent
Lambeth
Lewisham
Tower Hamlets
Kensington and ChelseaHammersmith and Fulham
Brent
Barnet
Barnet
Southwark
Tower Hamlets
Camden
Tower Hamlets
Waltham Forest
Newham
Islington
Richmond upon Thames
Lewisham
Camden
Westminster
Greenwich
Kensington and Chelsea
Barnet
Westminster
Lewisham
Waltham Forest
Hounslow, Ealing, Hammersmith and Fulham
Brent
Barnet
Lambeth, Wandsworth
Islington
Barnet
Merton
Barnet
Westminster
Barnet, Brent, Camden
Lewisham
Bexley
Haringey
Bromley
Tower Hamlets
Newham
Hackney
Dartford
Islington
Southwark
Lewisham
Brent
Southwark
Ealing
Kensington and Chelsea
Wandsworth
Southwark
Barnet
Newham
Richmond upon 

Sampling our data

In [ ]:
venues_in_London.head()

Neighbourhood  ...  Venue Category
0  Bexley, Greenwich   ...   Historic Site
1  Bexley, Greenwich   ...     Supermarket
2  Bexley, Greenwich   ...     Supermarket
3  Bexley, Greenwich   ...   Train Station
4  Bexley, Greenwich   ...     Coffee Shop

[5 rows x 5 columns]

In [ ]:
venues_in_London.shape

(10629, 5)

Wow, we have scraped together 10629 records for venues. This will definitely make the clustering interesting.

We need to now see how many Venue Categories are there for further processing

In [ ]:
venues_in_London.groupby('Venue Category').max()

Neighbourhood  ...                        Venue
Venue Category                               ...                             
Adult Boutique                    Islington  ...  Sh! Women's Erotic Emporium
African Restaurant              Westminster  ...           Red Sea Restaurant
American Restaurant          Waltham Forest  ...                  Spielburger
Antique Shop                    Westminster  ...     The London Silver Vaults
Arcade                          Westminster  ...           Novelty Automation
...                                     ...  ...                          ...
Wings Joint          Hammersmith and Fulham  ...                     Wingmans
Women's Store                   Westminster  ...           Vivien of Holloway
Xinjiang Restaurant               Southwark  ...                    Silk Road
Yoga Studio                     Westminster  ...                    yogahaven
Zoo Exhibit                          Camden  ...                Penguin Beach

[293 rows x 4 columns]

We can see 293 records, just goes to show how diverse and interesting the place is.



We need to Encode our venue categories to get a better result for our clustering

In [ ]:
London_venue_cat = pd.get_dummies(venues_in_London[['Venue Category']], prefix="", prefix_sep="")
London_venue_cat

Adult Boutique  African Restaurant  ...  Yoga Studio  Zoo Exhibit
0                   0                   0  ...            0            0
1                   0                   0  ...            0            0
2                   0                   0  ...            0            0
3                   0                   0  ...            0            0
4                   0                   0  ...            0            0
...               ...                 ...  ...          ...          ...
10624               0                   0  ...            0            0
10625               0                   0  ...            0            0
10626               0                   0  ...            0            0
10627               0                   0  ...            0            0
10628               0                   0  ...            0            0

[10629 rows x 293 columns]

Adding Neighbourhood into the mix.

In [ ]:
London_venue_cat['Neighbourhood'] = venues_in_London['Neighbourhood'] 

# moving neighborhood column to the first column
fixed_columns = [London_venue_cat.columns[-1]] + list(London_venue_cat.columns[:-1])
London_venue_cat = London_venue_cat[fixed_columns]

London_venue_cat.head()

Neighbourhood  Adult Boutique  ...  Yoga Studio  Zoo Exhibit
0  Bexley, Greenwich                0  ...            0            0
1  Bexley, Greenwich                0  ...            0            0
2  Bexley, Greenwich                0  ...            0            0
3  Bexley, Greenwich                0  ...            0            0
4  Bexley, Greenwich                0  ...            0            0

[5 rows x 294 columns]

We will group the Neighbourhoods and calculate the mean venue categories value in each Neighbourhood

In [ ]:
London_grouped = London_venue_cat.groupby('Neighbourhood').mean().reset_index()
London_grouped.head()

Neighbourhood  Adult Boutique  ...  Yoga Studio  Zoo Exhibit
0                 Barnet             0.0  ...          0.0          0.0
1  Barnet, Brent, Camden             0.0  ...          0.0          0.0
2                 Bexley             0.0  ...          0.0          0.0
3      Bexley, Greenwich             0.0  ...          0.0          0.0
4     Bexley, Greenwich              0.0  ...          0.0          0.0

[5 rows x 294 columns]

Let's make a function to get the top most common venue categories

In [ ]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

There are way too many venue categories, we can take the top 10 to cluster the neighbourhoods.<br>

Creating a function to label the columns of the venue correctly

In [ ]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighbourhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

Getting the top venue categories in London

In [ ]:
# create a new dataframe for London
neighborhoods_venues_sorted_london = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted_london['Neighbourhood'] = London_grouped['Neighbourhood']

for ind in np.arange(London_grouped.shape[0]):
    neighborhoods_venues_sorted_london.iloc[ind, 1:] = return_most_common_venues(London_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted_london.head()

Neighbourhood  ... 10th Most Common Venue
0                 Barnet  ...               Pharmacy
1  Barnet, Brent, Camden  ...    Filipino Restaurant
2                 Bexley  ...               Bus Stop
3      Bexley, Greenwich  ...     Falafel Restaurant
4     Bexley, Greenwich   ...         Farmers Market

[5 rows x 11 columns]

**4.5.2 Paris**<br>

Coordinates of paris

In [ ]:
paris = geocode(address='Paris, France, FR')[0]
paris_lng_coords = paris['location']['x']
paris_lat_coords = paris['location']['y']
print("The geolocation of Paris: ", paris_lat_coords, paris_lng_coords)

The geolocation of Paris:  48.85341000000005 2.3488000000000397


Vizualize the map of paris

In [ ]:
# Creating the map of Paris
map_Paris= folium.Map(location=[paris_lat_coords, paris_lng_coords], zoom_start=12)
map_Paris

# adding markers to map
for latitude, longitude, borough, town in zip(paris_combined_data['Latitude'], paris_combined_data['Longitude'], paris_combined_data['nom_comm'], paris_combined_data['nom_dept']):
    label = '{}, {}'.format(town, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [latitude, longitude],
        radius=5,
        popup=label,
        color='Blue',
        fill=True,
        fill_opacity=0.8
        ).add_to(map_Paris)  
    
map_Paris

*Venues in paris:*

Using our previously defined function. Let's get the neaby venues present in each neighbourhood of Paris



In [ ]:
venues_in_Paris = getNearbyVenues(paris_combined_data['nom_comm'], paris_combined_data['Latitude'], paris_combined_data['Longitude'])

PARIS-9E-ARRONDISSEMENT
PARIS-2E-ARRONDISSEMENT
PARIS-11E-ARRONDISSEMENT
PARIS-3E-ARRONDISSEMENT
PARIS-6E-ARRONDISSEMENT
PARIS-19E-ARRONDISSEMENT
PARIS-20E-ARRONDISSEMENT
PARIS-15E-ARRONDISSEMENT
PARIS-5E-ARRONDISSEMENT
PARIS-7E-ARRONDISSEMENT
PARIS-8E-ARRONDISSEMENT
PARIS-13E-ARRONDISSEMENT
PARIS-12E-ARRONDISSEMENT
PARIS-18E-ARRONDISSEMENT
PARIS-1ER-ARRONDISSEMENT
PARIS-10E-ARRONDISSEMENT
PARIS-16E-ARRONDISSEMENT
PARIS-4E-ARRONDISSEMENT
PARIS-17E-ARRONDISSEMENT
PARIS-14E-ARRONDISSEMENT


Sampling our Data

In [ ]:
venues_in_Paris.head()

Neighbourhood  ...                 Venue Category
0  PARIS-9E-ARRONDISSEMENT  ...                       Wine Bar
1  PARIS-9E-ARRONDISSEMENT  ...                          Plaza
2  PARIS-9E-ARRONDISSEMENT  ...                   Gourmet Shop
3  PARIS-9E-ARRONDISSEMENT  ...  Vegetarian / Vegan Restaurant
4  PARIS-9E-ARRONDISSEMENT  ...          Venezuelan Restaurant

[5 rows x 5 columns]

In [ ]:
venues_in_Paris.shape

(1306, 5)

We have managed to collect 1306 venue records for the neighbourhoods in Paris



We need to now see how many Venue Categories are there for further processing

In [ ]:
venues_in_Paris.groupby('Venue Category').max()

Neighbourhood  ...                                     Venue
Venue Category                                    ...                                          
Afghan Restaurant       PARIS-11E-ARRONDISSEMENT  ...                               Afghanistan
African Restaurant       PARIS-9E-ARRONDISSEMENT  ...                         Wally Le Saharien
American Restaurant     PARIS-19E-ARRONDISSEMENT  ...                                  Harper's
Antique Shop             PARIS-9E-ARRONDISSEMENT  ...                   Hôtel des Ventes Drouot
Argentinian Restaurant   PARIS-3E-ARRONDISSEMENT  ...                                     Anahi
...                                          ...  ...                                       ...
Wine Bar                 PARIS-9E-ARRONDISSEMENT  ...                          Vingt Vins d'Art
Wine Shop                PARIS-3E-ARRONDISSEMENT  ...                            Trois Fois Vin
Women's Store            PARIS-6E-ARRONDISSEMENT  ...                      L'Appartement Sézane
Zoo                     PARIS-12E-ARRONDISSEMENT  ...                  Parc zoologique de Paris
Zoo Exhibit             PARIS-12E-ARRONDISSEMENT  ...  Grande Serre du Parc Zoologique de Paris

[205 rows x 4 columns]

205 records in Paris. Paris is also multicultural and diverse as London

We need to Encode our venue categories to get a better result for our clustering

In [ ]:
Paris_venue_cat = pd.get_dummies(venues_in_Paris[['Venue Category']], prefix="", prefix_sep="")
Paris_venue_cat

Afghan Restaurant  African Restaurant  ...  Zoo  Zoo Exhibit
0                     0                   0  ...    0            0
1                     0                   0  ...    0            0
2                     0                   0  ...    0            0
3                     0                   0  ...    0            0
4                     0                   0  ...    0            0
...                 ...                 ...  ...  ...          ...
1301                  0                   0  ...    0            0
1302                  0                   0  ...    0            0
1303                  0                   0  ...    0            0
1304                  0                   0  ...    0            0
1305                  0                   0  ...    0            0

[1306 rows x 205 columns]

Adding Neighbourhoods to our Data

In [ ]:
Paris_venue_cat['Neighbourhood'] = venues_in_Paris['Neighbourhood'] 

# moving neighborhood column to the first column
fixed_columns = [Paris_venue_cat.columns[-1]] + list(Paris_venue_cat.columns[:-1])
Paris_venue_cat = Paris_venue_cat[fixed_columns]

Paris_venue_cat.head()

Neighbourhood  Afghan Restaurant  ...  Zoo  Zoo Exhibit
0  PARIS-9E-ARRONDISSEMENT                  0  ...    0            0
1  PARIS-9E-ARRONDISSEMENT                  0  ...    0            0
2  PARIS-9E-ARRONDISSEMENT                  0  ...    0            0
3  PARIS-9E-ARRONDISSEMENT                  0  ...    0            0
4  PARIS-9E-ARRONDISSEMENT                  0  ...    0            0

[5 rows x 206 columns]

We will group the Neighbourhoods and calculate the mean venue categories value in each Neighbourhood

In [ ]:
Paris_grouped = Paris_venue_cat.groupby('Neighbourhood').mean().reset_index()
Paris_grouped.head()

Neighbourhood  Afghan Restaurant  ...  Zoo  Zoo Exhibit
0  PARIS-10E-ARRONDISSEMENT            0.00000  ...  0.0          0.0
1  PARIS-11E-ARRONDISSEMENT            0.02381  ...  0.0          0.0
2  PARIS-12E-ARRONDISSEMENT            0.00000  ...  0.2          0.2
3  PARIS-13E-ARRONDISSEMENT            0.00000  ...  0.0          0.0
4  PARIS-14E-ARRONDISSEMENT            0.00000  ...  0.0          0.0

[5 rows x 206 columns]

Reusing our previously defined function to get the top venue categories in the neighbourhoods of Paris.

In [ ]:
# create a new dataframe for Paris
neighborhoods_venues_sorted_paris = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted_paris['Neighbourhood'] = Paris_grouped['Neighbourhood']

for ind in np.arange(Paris_grouped.shape[0]):
    neighborhoods_venues_sorted_paris.iloc[ind, 1:] = return_most_common_venues(Paris_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted_paris.head()

Neighbourhood  ... 10th Most Common Venue
0  PARIS-10E-ARRONDISSEMENT  ...      Indian Restaurant
1  PARIS-11E-ARRONDISSEMENT  ...      Afghan Restaurant
2  PARIS-12E-ARRONDISSEMENT  ...            Flower Shop
3  PARIS-13E-ARRONDISSEMENT  ...           Gourmet Shop
4  PARIS-14E-ARRONDISSEMENT  ...                  Plaza

[5 rows x 11 columns]

## **5. Model Building**<br>

### **5.1 Model Building for London neighbourhood**

Let's cluster the city of london to roughly 5 to make it easier to analyze.<br>

We use the K Means clustering technique to do so.

In [ ]:
k_num_clusters = 5

London_grouped_clustering = London_grouped.drop('Neighbourhood', 1)

# run k-means clustering
kmeans_london = KMeans(n_clusters=k_num_clusters, random_state=0).fit(London_grouped_clustering)
kmeans_london

KMeans(algorithm='auto', copy_x=True, init='k-means++', max_iter=300,
       n_clusters=5, n_init=10, n_jobs=None, precompute_distances='auto',
       random_state=0, tol=0.0001, verbose=0)

Labelling clustered data

In [ ]:
kmeans_london.labels_

array([0, 2, 2, 4, ..., 0, 0, 0, 0], dtype=int32)

So our model has labeled the city

In [ ]:
neighborhoods_venues_sorted_london.insert(0, 'Cluster Labels', kmeans_london.labels_ +1)

Join London_merged with our neighbourhood venues sorted to add latitude & longitude for each of the neighborhood to prepare it for plotting

In [ ]:
london_data = london_merged

london_data = london_data.join(neighborhoods_venues_sorted_london.set_index('Neighbourhood'), on='borough')

london_data.head()

borough  ... 10th Most Common Venue
0              Bexley, Greenwich   ...         Farmers Market
1  Ealing, Hammersmith and Fulham  ...             Food Court
6                            City  ...               Wine Bar
7                     Westminster  ...                 Bakery
9                         Bromley  ...            Golf Course

[5 rows x 16 columns]

Drop all the NaN values to prevent data skew

In [ ]:
london_data_nonan = london_data.dropna(subset=['Cluster Labels'])

Visualizing cluster neighbourhood:

Let's plot the clusters

In [ ]:
map_clusters_london = folium.Map(location=[london_lat_coords, london_lng_coords], zoom_start=12)

# set color scheme for the clusters
x = np.arange(k_num_clusters)
ys = [i + x + (i*x)**2 for i in range(k_num_clusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(london_data_nonan['latitude'], london_data_nonan['longitude'], london_data_nonan['borough'], london_data_nonan['Cluster Labels']):
    label = folium.Popup('Cluster ' + str(int(cluster) +1) + '\n' + str(poi) , parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[int(cluster-1)],
        fill=True,
        fill_color=rainbow[int(cluster-1)]
        ).add_to(map_clusters_london)
        
map_clusters_london

*Examining our clusters:*

Cluster 1:

In [ ]:
london_data_nonan.loc[london_data_nonan['Cluster Labels'] == 1, london_data_nonan.columns[[1] + list(range(5, london_data_nonan.shape[1]))]]

town  ...  10th Most Common Venue
6                    LONDON  ...                Wine Bar
7                    LONDON  ...                  Bakery
10                   LONDON  ...          Breakfast Spot
12                   LONDON  ...          Breakfast Spot
14           BARNET, LONDON  ...                Pharmacy
..                      ...  ...                     ...
523                  LONDON  ...      Seafood Restaurant
524  LONDON, WOODFORD GREEN  ...              Steakhouse
527                  LONDON  ...                Pharmacy
528                  LONDON  ...                    Park
530                  LONDON  ...       French Restaurant

[295 rows x 12 columns]

Cluster 2

In [ ]:
london_data_nonan.loc[london_data_nonan['Cluster Labels'] == 2, london_data_nonan.columns[[1] + list(range(5, london_data_nonan.shape[1]))]]

town  ...  10th Most Common Venue
379  HARROW, STANMOREEDGWARE, LONDON  ...             Event Space

[1 rows x 12 columns]

Cluster 3

In [ ]:
london_data_nonan.loc[london_data_nonan['Cluster Labels'] == 3, london_data_nonan.columns[[1] + list(range(5, london_data_nonan.shape[1]))]]

town  ...  10th Most Common Venue
0                 LONDON  ...          Farmers Market
9                 LONDON  ...             Golf Course
29     BECKENHAM, LONDON  ...             Golf Course
45   BEXLEYHEATH, LONDON  ...                Bus Stop
121               LONDON  ...     Filipino Restaurant
124               LONDON  ...                Bus Stop
127               LONDON  ...             Golf Course
292       LONDON, SIDCUP  ...                Bus Stop
317               LONDON  ...             Golf Course
361               LONDON  ...             Golf Course
507               LONDON  ...                Bus Stop

[11 rows x 12 columns]

Cluster 4

In [ ]:
london_data_nonan.loc[london_data_nonan['Cluster Labels'] == 4, london_data_nonan.columns[[1] + list(range(5, london_data_nonan.shape[1]))]]

town  Cluster Labels  ... 9th Most Common Venue 10th Most Common Venue
1  LONDON               4  ...            Food Stand             Food Court

[1 rows x 12 columns]

Cluster 5

In [ ]:
london_data_nonan.loc[london_data_nonan['Cluster Labels'] == 5, london_data_nonan.columns[[1] + list(range(5, london_data_nonan.shape[1]))]]

town  ...  10th Most Common Venue
168  LONDON, WELLING  ...      Falafel Restaurant
459    LONDON, ERITH  ...      Falafel Restaurant

[2 rows x 12 columns]

### **5.2 Modelling for Paris Neighbourhood**<br>

Let's cluster the city of Paris to roughly 5 to make it easier to analyze.

As previously we did, K Means clustering technique is used to do so.

In [ ]:
# set number of clusters
k_num_clusters = 5

Paris_grouped_clustering = Paris_grouped.drop('Neighbourhood', 1)

# run k-means clustering
kmeans_Paris = KMeans(n_clusters=k_num_clusters, random_state=0).fit(Paris_grouped_clustering)
kmeans_Paris

KMeans(algorithm='auto', copy_x=True, init='k-means++', max_iter=300,
       n_clusters=5, n_init=10, n_jobs=None, precompute_distances='auto',
       random_state=0, tol=0.0001, verbose=0)

Labelling clustured data

In [ ]:
kmeans_Paris.labels_

array([2, 2, 3, 4, 1, 2, 0, 1, 2, 2, 2, 2, 2, 2, 2, 2, 2, 1, 1, 2], dtype=int32)

So our model has labeled the city, we insert it in our data.

In [ ]:
neighborhoods_venues_sorted_paris.insert(0, 'Cluster Labels', kmeans_Paris.labels_ +1)

Join paris_combined_data with our neighbourhood venues sorted to add latitude & longitude for each of the neighborhood to prepare it for plotting

In [ ]:
paris_data = paris_combined_data

paris_data = paris_data.join(neighborhoods_venues_sorted_paris.set_index('Neighbourhood'), on='nom_comm')

paris_data.head()

postal_code  ... 10th Most Common Venue
0       75009  ...   Gym / Fitness Center
1       75002  ...                  Hotel
2       75011  ...      Afghan Restaurant
3       75003  ...         Sandwich Place
4       75006  ...             Restaurant

[5 rows x 16 columns]

Drop all the NaN values to prevent data skew

In [ ]:
paris_data_nonan = paris_data.dropna(subset=['Cluster Labels'])

Let's plot the clusters


In [ ]:
map_clusters_paris = folium.Map(location=[paris_lat_coords, paris_lng_coords], zoom_start=12)

# set color scheme for the clusters
x = np.arange(k_num_clusters)
ys = [i + x + (i*x)**2 for i in range(k_num_clusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(paris_data_nonan['Latitude'], paris_data_nonan['Longitude'], paris_data_nonan['nom_comm'], paris_data_nonan['Cluster Labels']):
    label = folium.Popup('Cluster ' + str(int(cluster) +1) + ' ' + str(poi) , parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[int(cluster-1)],
        fill=True,
        fill_color=rainbow[int(cluster-1)],
        fill_opacity=0.8
        ).add_to(map_clusters_paris)
        
map_clusters_paris

*Examining our clusters:*

Cluster 1

In [ ]:
paris_data_nonan.loc[paris_data_nonan['Cluster Labels'] == 1, paris_data_nonan.columns[[1] + list(range(5, paris_data_nonan.shape[1]))]]

nom_comm  ...  10th Most Common Venue
16  PARIS-16E-ARRONDISSEMENT  ...    Ethiopian Restaurant

[1 rows x 12 columns]

Cluster 2

In [ ]:
paris_data_nonan.loc[paris_data_nonan['Cluster Labels'] == 2, paris_data_nonan.columns[[1] + list(range(5, paris_data_nonan.shape[1]))]]

nom_comm  ...  10th Most Common Venue
9    PARIS-7E-ARRONDISSEMENT  ...                  Bistro
10   PARIS-8E-ARRONDISSEMENT  ...                    Park
18  PARIS-17E-ARRONDISSEMENT  ...            Burger Joint
19  PARIS-14E-ARRONDISSEMENT  ...                   Plaza

[4 rows x 12 columns]

Cluster 3

In [ ]:
paris_data_nonan.loc[paris_data_nonan['Cluster Labels'] == 3, paris_data_nonan.columns[[1] + list(range(5, paris_data_nonan.shape[1]))]]

nom_comm  ...  10th Most Common Venue
0    PARIS-9E-ARRONDISSEMENT  ...    Gym / Fitness Center
1    PARIS-2E-ARRONDISSEMENT  ...                   Hotel
2   PARIS-11E-ARRONDISSEMENT  ...       Afghan Restaurant
3    PARIS-3E-ARRONDISSEMENT  ...          Sandwich Place
4    PARIS-6E-ARRONDISSEMENT  ...              Restaurant
5   PARIS-19E-ARRONDISSEMENT  ...              Steakhouse
6   PARIS-20E-ARRONDISSEMENT  ...                    Café
7   PARIS-15E-ARRONDISSEMENT  ...                    Park
8    PARIS-5E-ARRONDISSEMENT  ...          Science Museum
13  PARIS-18E-ARRONDISSEMENT  ...                    Café
14  PARIS-1ER-ARRONDISSEMENT  ...                  Bakery
15  PARIS-10E-ARRONDISSEMENT  ...       Indian Restaurant
17   PARIS-4E-ARRONDISSEMENT  ...                   Plaza

[13 rows x 12 columns]

Cluster 4

In [ ]:
paris_data_nonan.loc[paris_data_nonan['Cluster Labels'] == 4, paris_data_nonan.columns[[1] + list(range(5, paris_data_nonan.shape[1]))]]

nom_comm  ...  10th Most Common Venue
12  PARIS-12E-ARRONDISSEMENT  ...             Flower Shop

[1 rows x 12 columns]

Cluster 5

In [ ]:
paris_data_nonan.loc[paris_data_nonan['Cluster Labels'] == 5, paris_data_nonan.columns[[1] + list(range(5, paris_data_nonan.shape[1]))]]

nom_comm  ...  10th Most Common Venue
11  PARIS-13E-ARRONDISSEMENT  ...            Gourmet Shop

[1 rows x 12 columns]

## **6. Results and Discussion**<br>

The neighbourhoods of London are very mulitcultural. There are a lot of different cusines including Indian, Italian, Turkish and Chinese. London seems to take a step further in this direction by having a lot of Restaurants, bars, juice bars, coffee shops, Fish and Chips shop and Breakfast spots. It has a lot of shopping options too with that of the Flea markets, flower shops, fish markets, Fishing stores, clothing stores. The main modes of transport seem to be Buses and trains. For leisure, the neighbourhoods are set up to have lots of parks, golf courses, zoo, gyms and Historic sites.<br>

Overall, the city of London offers a multicultural, diverse and certainly an entertaining experience.<br>

Paris is relatively small in size geographically. It has a wide variety of cusines and eateries including French, Thai, Cambodian, Asian, Chinese etc. There are a lot of hangout spots including many Restaurants and Bars. Paris has a lot of Bistro's. Different means of public transport in Paris which includes buses, bikes, boats or ferries. For leisure and sight seeing, there are a lot of Plazas, Trails, Parks, Historic sites, clothing shops, Art galleries and Museums. Overall, Paris seems like the relaxing vacation spot with a mix of lakes, historic spots and a wide variety of cusines to try out.

## **7. Conclusion**<br>

The purpose of this project was to explore the cities of London and Paris and see how attractive it is to potential tourists and migrants. We explored both the cities based on their postal codes and then extrapolated the common venues present in each of the neighbourhoods finally concluding with clustering similar neighbourhoods together.

We could see that each of the neighbourhoods in both the cities have a wide variety of experiences to offer which is unique in it's own way. The cultural diversity is quite evident which also gives the feeling of a sense of inclusion.

Both Paris and London seem to offer a vacation stay or a romantic gateaway with a lot of places to explore, beautiful landscapes and a wide variety of culture.Overall, it's upto the stakeholders to decide which experience they would prefer more and which would more to their liking.

## **References**<br>

1. The Battle of Neighbourhood — My London’s Perspective by Dayo John <br>
https://medium.com/@oludayo.oguntoyinbo/the-battle-of-neighbourhood-my-londons-perspective-d363163771e0

2. The Battle of neighborhoods! What is the best place where can I start my restaurant business in Paris? by Zakaria BOUZIANE<br>
https://www.linkedin.com/pulse/battle-neighborhoods-what-best-place-where-can-i-start-bouziane/

3. Foursquare API <br>
https://foursquare.com/

4. ArcGIS API<br>
https://www.arcgis.com/index.html

5. A tale of two cities storytelling with data<br>
https://towardsdatascience.com/a-tale-of-two-cities-e693c15b3ddb

6. A tale of two cities storytelling with data by Warwick Rommelrath<br>
https://towardsdatascience.com/a-framework-for-running-data-science-projects-fd37b26a4389